<a href="https://colab.research.google.com/github/660510762-Chidsanupong/660510762-Chidsanupong/blob/main/Lab04_Naive_Bayes_Grid_and_Random_Search_660510762.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Statistical Learning for Data Science 2 (229352)
#### Instructor: Donlapark Ponnoprat

#### [Course website](https://donlapark.pages.dev/229352/)

## Lab #4

## 660510762 ชิษณุพงศ์ การะเกตุ

In [1]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import classification_report
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline

from scipy.stats import uniform

In [2]:
train = fetch_20newsgroups(subset='train')
test = fetch_20newsgroups(subset='test')

Xtrain = train.data[:3000]
ytrain = train.target[:3000]
Xtest = test.data[:500]
ytest = test.target[:500]

print("X:", len(Xtest))
print("y:", len(ytest))

X: 500
y: 500


### Naive Bayes [(Documentation)](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html)

In [3]:
# Create a pipeline with TF-IDF vectorizer and Multinomial Naive Bayes classifier
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer()),
    ('mnb', MultinomialNB())
])

# Define the parameter grid for alpha
param_grid = {'mnb__alpha': [0.01, 0.1, 0.5, 1.0, 2.0]}

# Perform GridSearchCV
grid_search = GridSearchCV(pipeline, param_grid, cv=5, scoring='f1_macro')
grid_search.fit(Xtrain, ytrain)

# Print the best alpha and the corresponding f1_macro score on the test set
print("Best alpha:", grid_search.best_params_['mnb__alpha'])
print("f1_macro score on test set:", grid_search.score(Xtest, ytest))

Best alpha: 0.01
f1_macro score on test set: 0.7725871193182322


### Random Search Cross-Validation [(Documentation)](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html)

### Uniform distribution in `Scipy` [(Documentation)](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.uniform.html)

In [4]:
# Define the parameter distribution for alpha using a uniform distribution
param_dist = {'mnb__alpha': uniform(loc=0, scale=2.0)} # Using scale=2.0 as an example range

# Perform RandomizedSearchCV
random_search = RandomizedSearchCV(pipeline, param_dist, n_iter=10, cv=5, scoring='f1_macro', random_state=42)
random_search.fit(Xtrain, ytrain)

# Print the best alpha and the corresponding f1_macro score on the test set
print("Best alpha:", random_search.best_params_['mnb__alpha'])
print("f1_macro score on test set:", random_search.score(Xtest, ytest))

Best alpha: 0.11616722433639892
f1_macro score on test set: 0.7103431960621382


#### Exercise

1. For the Naive Bayes model, use grid search 5-fold cross-validation across different values of `alpha` to find the best model.

2. For the best value of `alpha`, compute the `f1_macro` score on the test set.
* What value of `alpha` did you obtain?
* What is the model's `f1_macro` score?

3. Repeat Exercise 1 and 2 for **random search** 5-fold cross validation across different values of `alpha`. Compute the `f1_macro` score on the test set.
* What value of `alpha` did you obtain?
* Did you get a better `f1_macro` score compared to grid search in Exercise 2?

In [5]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.metrics import f1_score
from scipy.stats import uniform

# สร้าง Pipeline
pipe = Pipeline([
    ('tfidf', TfidfVectorizer()),
    ('mnb', MultinomialNB())
])

# --- Exercise 1: GridSearchCV ---
param_grid = {'mnb__alpha': [0.01, 0.1, 0.5, 1.0, 2.0]}
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='f1_macro')
grid.fit(Xtrain, ytrain)

# ทำนายและคำนวณ f1_macro บน test set
y_pred_grid = grid.predict(Xtest)
f1_grid = f1_score(ytest, y_pred_grid, average='macro')

print("Grid Search")
print("Best alpha:", grid.best_params_['mnb__alpha'])
print("f1_macro score on test set:", round(f1_grid, 4))

# --- Exercise 3: RandomizedSearchCV ---
param_dist = {'mnb__alpha': uniform(0, 2.0)}  # ค่า alpha จาก 0 ถึง 2
random = RandomizedSearchCV(pipe, param_dist, n_iter=10, cv=5, scoring='f1_macro', random_state=42)
random.fit(Xtrain, ytrain)

# ทำนายและคำนวณ f1_macro บน test set
y_pred_random = random.predict(Xtest)
f1_random = f1_score(ytest, y_pred_random, average='macro')

print("\n Random Search")
print("Best alpha:", round(random.best_params_['mnb__alpha'], 4))
print("f1_macro score on test set:", round(f1_random, 4))

Grid Search
Best alpha: 0.01
f1_macro score on test set: 0.7726

 Random Search
Best alpha: 0.1162
f1_macro score on test set: 0.7103


## สรุปผล
Grid Search ได้ alpha = 0.01 และได้ค่า f1_macro = 0.7726

Random Search ได้ alpha ≈ 0.1162 และ f1_macro = 0.7103

Grid Search ให้ผลลัพธ์ดีกว่า